totoify_grafeas/totoifylib.py

```diff
@@ -0,0 +1,269 @@
+      else:
+        byproducts["custom_values"][key] = value
+
+    environment = {"custom_values": {}}
+    for key, value in link.environment.items():
+      environment["custom_values"][key] = value
+
+    return cls(materials=materials, products=products, command=link.command,
+               byproducts=byproducts, environment=environment)
+
+  def to_link(self, step_name):
+    """Performs conversion from GrafeasLink to in_toto.models.link.Link"""
+    materials = {}
+    products = {}
+    command = []
+    byproducts = {}
+    environment = {}
+
+    for item in self.materials:
+      materials[item["resource_uri"]] = item["hashes"]
+
+    for item in self.products:
+      products[item["resource_uri"]] = item["hashes"]
+
+    command = self.command
+
+    for key, value in self.byproducts.items():
+      if key == "custom_values":
+        continue
+      byproducts[key] = value
+
+    if "custom_values" in self.byproducts:
+      for key, value in \
+          self.byproducts["custom_values"].items():
+        if key == "return-value":
+          # This highlights a special case - in-toto's reference implementations
+          # store return value as an integer while Grafeas allows only strings
+          byproducts[key] = int(value)
+        else:
+          byproducts[key] = value
+
+    for key, value in self.environment.items():
+      if key == "custom_values":
+        continue
+      environment[key] = value
+
+    if "custom_values" in self.environment:
+      for key, value in \
+          self.environment["custom_values"].items():
+        environment[key] = value
+
+    return Link(name=step_name,
+                materials=materials,
+                products=products,
+                byproducts=byproducts,
+                command=command,
+                environment=environment)
+
+
+  def __repr__(self):
+    return json.dumps(self.to_dict())
+
+  def to_dict(self):
+    return {
+        "command": self.command,
+        "materials": self.materials,
+        "products": self.products,
+        "environment": self.environment,
+        "byproducts": self.byproducts
+      }
+
+
+class GrafeasInTotoOccurrence:
+  """Wrapper to hold a Grafeas occurrence complete with in-toto metadata.
+
+  Provides methods to dump occurrence fields into a json string, and to load
+  the occurrence file into the class.
+
+  Attributes:
+    note_name: A string of the name of the note.
+    kind: Set to the string "INTOTO" to indicate the type of occurrence.
+    resource: A dict of resource uri.
+    intoto: A dictionary with key "signatures", a list of signatures consisting
+        of key id and signature hash, and key "signed", the GrafeasLink class.
+  """
+
+  def __init__(self, intoto=None, note_name=None, resource_uri=None):
+    """Initalizes intoto "signed" fields with a GrafeasLink."""
+    if intoto is None:
+      self.intoto = {
+          "signatures": [],
+          "signed": None
```

Comment:
I'm using `GrafeasLink()` now. As for default arguments, I think not. The intended workflow for use is `.load`/`.to_link`/`.from_link` etc. Should this move to just compulsory arguments?

Reply:
If you only intend to use .load, .from_link and those methods only pass non-None values for all constructor parameters then I think you can make them compulsory. This would also spare you all of the `if <passed_arg> is None: ... / else: ...` boilerplate  code. :)